src/org/python/core/JyAttribute.java

```diff
@@ -197,18 +193,51 @@ public static boolean hasAttr(PyObject ob, byte attr_type) {
      * Retrieves the attribute of the given type from the given
      * {@link org.python.core.PyObject}.
      * If no attribute of the given type is attached, null is returned.
+     * 
+     * This class is generally thread safe, but this method is not synchronized with
+     * concurrent write access to avoid performance issues for multithreaded scenarios.
+     * The trade-off for that choice is that in case of concurrent write access, it is
+     * unspecified whether this method returns according to the old state or the new state.
+     * This method is thread safe in the sense that a concurrent write access never exposes
+     * an invalid intermediate state to this read access.
      */
     public static Object getAttr(PyObject ob, byte attr_type) {
-        synchronized (ob) {
-            if (ob.attributes instanceof JyAttribute) {
-                JyAttribute att = (JyAttribute) ob.attributes;
-                while (att != null && att.attr_type < attr_type) {
-                    att = att.getNext();
-                }
-                return att != null && att.attr_type == attr_type ? att.getValue() : null;
+        /*
+        Here used to be a synchronized(ob) block encapsulating the whole method code.
+        It has been removed due to performance issues reported in
+        github.com/jython/jython/issues/360.
+        By making ob.attributes in PyObject and next and value in AttributeLink and
+        TransientAttributeLink volatile, the read access in this method does not
+        fatally interfere with a concurrent write operation.
+        (Write operations continue to be synchronized against each other.)
+        https://docs.oracle.com/javase/specs/jls/se17/html/jls-17.html#jls-17.4.4 states:
+        "A write to a volatile variable v (§8.3.1.4) synchronizes-with all subsequent reads
+         of v by any thread"
+        This induces an x happens-before y relation hb(x, y) according to
+        https://docs.oracle.com/javase/specs/jls/se17/html/jls-17.html#jls-17.4.5:
+        "If an action x synchronizes-with a following action y, then we also have hb(x, y)."
+        Also see https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html,
+        which confirms the anticipated behavior of volatile fields for Java 5 and onwards.
+        Write accesses in the linked attribute list are designed to update the list tail
```

Comment:
According to the specs, I think it would (volatile save me, that is). It should prevent reordering `newAtt.setNext` (because all it does is assign the arg to the volatile field `next`) with `ob.attributes` assignment. If a node in the middle is inserted, the reader might either be still before that node or have already passed it. Therefore, it is unspecified whether according to the old or new state is returned. The approach should, however, ensure that in no case unrelated attributes are temporarily unreachable, because the tail is already attached to the new middle node before it is inserted with `volatile` preventing reordering.

Reply:
The passage (JLS 17.4.5) that nags me a bit is:

>  if two actions share a happens-before relationship, they do not necessarily have to appear to have happened in that order to any code with which they do not share a happens-before relationship. Writes in one thread that are in a data race with reads in another thread may, for example, appear to occur out of order to those reads.

but I think you are right that because `newAtt.next = att` and `ob.attributes = newAtt` are both (inter-thread) actions thanks to `volatile`, the order of writes in the writer and the reader must be the same. Similarly, the order of reads in the reader must be the same to both threads. So if a read of `ob.attributes` yields `newAtt`, rather than a previous value, then `newAtt.next` must by then be `att` as desired.

What about `newAtt.attr_type`: might it still have its default value `0` because `newAtt.attr_type = attr_type` was reordered after the other writes? Accoding to Pugh, we are not guaranteed that initialisation of `newAtt` has completed.

Reply:
Re `attr_type`, I somehow thought that making it `final` would ensure high priority on initialization. Like that `final` would imply `volatile` semantics. If there is a point in making a field `final volatile` (and if that's even possible), then sure, we can add the volatile-bit to the `attr_type` field.

Reply:
From [JLS 17.5](https://docs.oracle.com/javase/specs/jls/se17/html/jls-17.html#jls-17.5):
> An object is considered to be completely initialized when its constructor finishes. A thread that can only see a reference to an object after that object has been completely initialized is guaranteed to see the correctly initialized values for that object's final fields.

Due to `volatile` preventing reordering, the constructor of `AttributeLink` should have finished before its assignment to `next` and therefore, the final field should be properly initialized.